src/FS.Embedded.Manifest.Task/Entry.cs

```diff
@@ -0,0 +1,120 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Extensions.FileProviders.Embedded.Manifest.Task.Internal
+{
+    /// <summary>
+    /// This type is for internal uses only and is not meant to be consumed by any other library.
+    /// </summary>
+    [DebuggerDisplay("{Name,nq}")]
+    public class Entry : IEquatable<Entry>
+    {
+        public bool IsFile { get; private set; }
+
+        public string Name { get; private set; }
+
+        public string AssemblyResourceName { get; private set; }
+
+        public ISet<Entry> Children { get; } = new SortedSet<Entry>(NameComparer.Instance);
+
+        public static Entry Directory(string name) =>
+            new Entry { Name = name };
+
+        public static Entry File(string name, string assemblyResourceName) =>
+            new Entry { Name = name, AssemblyResourceName = assemblyResourceName, IsFile = true };
+
+        internal void AddChild(Entry child)
+        {
+            if (IsFile)
+            {
+                throw new InvalidOperationException("Tried to add children to a file.");
+            }
+
+            if (Children.Contains(child))
+            {
+                throw new InvalidOperationException($"An item with the name '{child.Name}' already exists.");
+            }
+
+            Children.Add(child);
+        }
+
+        internal Entry GetDirectory(string currentSegment)
+        {
+            if (IsFile)
+            {
+                throw new InvalidOperationException("Tried to get a directory from a file.");
+            }
+
+            foreach (var child in Children)
+            {
+                if (child.HasName(currentSegment))
+                {
+                    if (child.IsFile)
+                    {
+                        throw new InvalidOperationException("Tried to find a directory but found a file instead");
+                    }
+                    else
+                    {
+                        return child;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private bool HasName(string currentSegment)
+        {
+            return string.Equals(Name, currentSegment, StringComparison.Ordinal);
+        }
+
+        public bool Equals(Entry other)
+        {
+            if (other == null || !other.HasName(Name) || other.IsFile != IsFile)
+            {
+                return false;
+            }
+
+            if (IsFile)
+            {
+                return string.Equals(other.AssemblyResourceName, AssemblyResourceName, StringComparison.Ordinal);
+            }
+            else
+            {
+                return SameChildren(Children, other.Children);
+            }
+        }
+
+        private bool SameChildren(ISet<Entry> left, ISet<Entry> right)
+        {
+            if (left.Count != right.Count)
```

Comment:
Isn't there a SetEquals on set?

Reply:
Equality for these sets is not based solely on the name. We check that we aren't adding two files with the same name but different metadata

Reply:
Yup. But you should be able to use ``SortedSet<Entry>.CreateSetComparer(EqualityComparer<Entry>.Default).Equals(left, right);`